bigtable/google/cloud/bigtable/table.py

```diff
@@ -363,6 +375,74 @@ def sample_row_keys(self):
         return response_iterator
 
 
+class _MutateRowsRetryableError(Exception):
+    """A retryable error in Mutate Rows response."""
+    pass
+
+
+class _RetryableMutateRowsWorker(object):
+    # pylint: disable=unsubscriptable-object
+    RETRY_CODES = (
+        StatusCode.DEADLINE_EXCEEDED.value[0],
+        StatusCode.ABORTED.value[0],
+        StatusCode.UNAVAILABLE.value[0],
+    )
+
+    def __init__(self, client, table_name, rows):
+        self.client = client
+        self.table_name = table_name
+        self.rows = rows
+        self.responses_statuses = [
+            None for _ in six.moves.xrange(len(self.rows))]
+
+    def __call__(self):
+        return self._do_mutate_retryable_rows()
+
+    def _is_retryable(self, status):  # pylint: disable=no-self-use
+        return (status is None or
+                status.code in _RetryableMutateRowsWorker.RETRY_CODES)
+
+    def _next_retryable_row_index(self, begin_index):
+        i = begin_index
+        while i < len(self.responses_statuses):
+            status = self.responses_statuses[i]
+            if self._is_retryable(status):
+                return i
+            i += 1
+        return -1
+
+    def _do_mutate_retryable_rows(self):
+        retryable_rows = []
+        for i, status in enumerate(self.responses_statuses):
+            if self._is_retryable(status):
+                retryable_rows.append(self.rows[i])
+
+        if not retryable_rows:
+            # All mutations are either successful or non-retryable now.
+            return self.responses_statuses
+
+        mutate_rows_request = _mutate_rows_request(
+            self.table_name, retryable_rows)
+        responses = self.client._data_stub.MutateRows(
+            mutate_rows_request)
+
+        num_retryable_responses = 0
+        for response in responses:
+            index = 0
+            for entry in response.entries:
+                index = self._next_retryable_row_index(index)
+                self.responses_statuses[index] = entry.status
+                if self._is_retryable(entry.status):
+                    num_retryable_responses += 1
+                if entry.status.code == 0:
+                    self.rows[index].clear()
+                index += 1
+
+        if num_retryable_responses:
+            raise _MutateRowsRetryableError()
```

Comment:
Would be good to include a list of rows & their failures in the exception

Reply:
I will include a list of `MutateRowsResponse.Entry` (with actual index to all rows). The caller already has a reference to the rows, so it can do any inspection.

Reply:
Done